include/seqan3/argument_parser/validators.hpp

```diff
@@ -297,68 +285,350 @@ class file_ext_validator
     //!\endcond
     void operator()(range_type const & v) const
     {
-        std::for_each(v.begin(), v.end(), [&] (auto cmp) { (*this)(cmp); });
+         std::for_each(v.begin(), v.end(), [&] (auto cmp) { this->operator()(cmp); });
+    }
+
+protected:
+
+    /*!\brief Validates the given filename path based on the specified extensions.
+     * \param path The filename path.
+     * \throws parser_invalid_argument if the specified extensions don't match the given path.
+     */
+    void validate_filename_impl(std::filesystem::path const & path) const
+    {
+        // If no valid extensions are given we can safely return here.
+        if (extensions.empty())
+            return;
+
+        // Check if extension is available.
+        if (!path.has_extension())
+            throw parser_invalid_argument{detail::to_string("The given filename ", path.string(),
+                                                            " has no extension. Expected one of the following valid"
+                                                            " extensions:", extensions, "!")};
+
+        // Drop the dot.
+        std::string tmp_str = path.extension().string();
+        auto drop_less_ext = tmp_str | view::drop(1);
+
+        // Compares the extensions in lower case.
+        auto cmp_lambda = [&] (std::string const & cmp)
+        {
+            return std::ranges::equal(cmp | view::to_lower, drop_less_ext | view::to_lower);
+        };
+
+        // Check if requested extension is present.
+        if (std::ranges::find_if(extensions, cmp_lambda) == extensions.end())
+        {
+            throw parser_invalid_argument{detail::to_string("Expected one of the following valid extensions: ",
+                                                             extensions, "! Got ", drop_less_ext, " instead!")};
+        }
+    }
+
+    /*!\brief Checks if the given path is readable.
+     * \param path The path to check.
+     * \returns `true` if readable, otherwise `false`.
+     */
+    bool path_is_readable(std::filesystem::path const & path) const
+    {
+        // Check if input directory is a readable.
+        if (std::filesystem::is_directory(path))
+        {
+            std::error_code ec{};
+            std::filesystem::directory_iterator{path, ec};  // if directory iterator cannot be created, ec will be set.
+            return !static_cast<bool>(ec); // test the error code.
+        }
+
+        // Must be a regular file.
+        if (!std::filesystem::is_regular_file(path))
+            throw std::invalid_argument{detail::to_string("Expected a regular file!", path, "!")};
+
+        std::ifstream file{path};
+        return file.is_open() && file.good();
+    }
```

Comment:
Hm, I don't think it's a good idea that we have different return paths from this function, i.e. that we return `false` for some files that are not readable and throw an exception for others.

I would either call it `void validate_readability` and always throw when something is wrong OR call it `bool is readable` and always return true/false and never throw. It makes it a lot easier to reason about.

Reply:
Since everything here seems to operate via exceptions, one might as well use the mechanism everywhere.

Reply:
good point

Reply:
Hm, now it is called `is_...` but it is void which is not really nice. I mean it's not super-important but I think in general the functions should be verbs and if it is actually called `is_foo` than it should return bool.
Why not `void validate_readability` ?

Reply:
I change this as well. No hard feelings.